dbms/src/Flash/Coprocessor/StreamWriter.h

```diff
@@ -35,12 +35,36 @@ class MPPDataPacket;
 
 namespace DB
 {
-struct StreamWriter
+
+struct CopStreamWriter
+{
+    grpc::ServerWriter<coprocessor::Response> * writer;
+    std::mutex write_mutex;
+
+    explicit CopStreamWriter(grpc::ServerWriter<coprocessor::Response> * writer_)
+        : writer(writer_)
+    {}
+    void write(tipb::SelectResponse & response)
+    {
+        coprocessor::Response resp;
+        if (!response.SerializeToString(resp.mutable_data()))
+            throw Exception("[StreamWriter]Fail to serialize response, response size: " + std::to_string(response.ByteSizeLong()));
+
+        GET_METRIC(tiflash_coprocessor_response_bytes, type_cop_stream).Increment(resp.ByteSizeLong());
+
+        std::lock_guard lk(write_mutex);
```

Comment:
It looks like the lock is useless here, because there is one and only one thread that will call write?

Reply:
The cop stream may have multiple threads.

Reply:
https://github.com/pingcap/tiflash/pull/7805/files#diff-79c3bf18678bcb4ec9ee47cb7f2f32232b4e52fae5a8ca3e00b244a1bbe780f6R149
The lambda here will only be called by a single thread

Reply:
But I think it's okay to keep the lock, because batchCop also has locks.
Although this lock is useless...